add hqDefine and hqImport, use for COMMCAREHQ.app_manager

dannyroberts opened the pull request:
@dimagi/js-team what do you think?

dannyroberts committed:
add hqDefine and hqImport, use for COMMCAREHQ.app_manager

millerdev commented on corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js:
```diff
@@ -0,0 +1,24 @@
+var COMMCAREHQ_MODULES = {};
+
+function hqDefine(path, moduleObject) {
```
How hard would it be make this signature match requirejs' `define` a bit more closely?

```js
function hqDefine(path, moduleFunc) {
```
`hqDefine` should call `moduleFunc` and the returned object is the module object. Alternately, we could make it take a function or object, and if it's a function call it and use the returned object, otherwise use the object as the module.

dannyroberts replied:
Thanks @millerdev that's a good idea. I think it's more in line javascript conventions anyway, so I'll change it to be a function.

millerdev commented on corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js:
```diff
@@ -0,0 +1,24 @@
+        }
+        object = object[parts[i]];
+    }
+    object[parts[i]] = moduleObject;
```
Am I understanding correctly that this will overwrite an already-defined module if two modules are defined with the same path? Seems like that could happen pretty easily if someone copy/pasted a module from one file to another and then forgot to update the path. Should we check for that and throw an error if it happens?

dannyroberts replied:
Yeah great point. I think something that's coming out of this is that even though we won't get all the fancy static analysis etc. that a real module system might promise, even just doing this much gives us the opportunity to add in some runtime checks that prevent or discourage bad behavior. I'll go ahead and add this check. Thanks!

millerdev commented:
I like it

esoergel commented:
Looks good!  This enforces the code separation benefits of a module system, even if it doesn't handle dependency management for you.  If/when we *do* get system.js or require.js or whatever working, modules defined in this pattern would be very straightforward to convert.
:star2:

czue commented:
would be cool to see some docs for this

dannyroberts committed:
Merge branch 'master' into hqImport

dannyroberts committed:
have hqDefine accept a module generator function

instead of the module itself. Makes it the norm to build the module within an enclosed scope

benrudolph commented:
cool this is pretty elegant

dannyroberts committed:
throw error if hqDefine called twice with same module path

dannyroberts committed:
add documentation comment to hqModules

dannyroberts commented:
@czue good call. Since it's still pretty early I decided to go with a half-way measure of adding a comment at the top of hqModules.js with a basic description of what it is and isn't, potential gotchas, and examples of how it should be used.

millerdev commented on corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js:
```diff
@@ -0,0 +1,66 @@
+function hqDefine(path, moduleAccessor) {
+    var parts = ['COMMCAREHQ_MODULES'].concat(path.split('/'));
+    var i;
+    var object = window;
```
tiny optimization
```js
var parts = path.split('/');
var object = COMMCAREHQ_MODULES;
```

dannyroberts replied:
fair enough. Fixed

millerdev commented on corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js:
```diff
@@ -0,0 +1,66 @@
+ * only a module *referencer*: "importing" a module doesn't automatically load it as
+ * a script to the page; it must already have been loaded with an explict script tag.
+ *
+ * Modules MUST have a name, and SHOULD be given then name
```
very minor:

s/then name/the name/

also, inconsistent wrapping with previous paragraph

dannyroberts committed:
simplify hqModules code slightly

dannyroberts committed:
fix comment typo and formatting

millerdev commented on corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js:
```diff
@@ -0,0 +1,65 @@
+        if (typeof object[parts[i]] === 'undefined') {
+            object[parts[i]] = {};
+        }
+        object = object[parts[i]];
```
Just noticed this now, and I don't really think it's a problem, just pointing it out: splitting the path and creating intermediate "package" objects containing nested packages or modules doesn't really provide any value here, does it? It could more simply do `COMMCAREHQ_MODULES[path] = module` (after the collision check) and nothing would break, right?

The only advantage to this setup that I can think of is runtime inspection by a human in a debugger, but even that seems of ~little value given that it will probably be almost as easy to look through a list of paths rather than navigating a hierarchical structure of objects to find a module.

No action needed, just thinking out loud.

dannyroberts committed:
simplify hqModules.js

to use the whole path a single key

dannyroberts committed:
throw nicer error if hqImport module undefined

dannyroberts commented:
@millerdev thanks for your great feedback. I've gone through and addressed them all, and I think the PR's much better for it!

millerdev commented:
:balloon: :tada:

biyeun commented:
YAY

proteusvacuum commented:
Very cool! Saw this referenced in the js reportout. I think @dimagi/team-commcare-hq should take a look too :100:

proteusvacuum commented on corehq/apps/hqwebapp/static/hqwebapp/js/hqModules.js:
```diff
@@ -0,0 +1,54 @@
+/*
+ * hqModules provides a poor man's module system for js. It is not a module *loader*,
```
did you come up with this? would probably make a great blog post.

orangejenny replied:
agree!

benrudolph replied:
:memo:

kaapstorm commented:
Sweet like :candy:
